### ISSUE TITLE
Rust 2018 & MSRV bump 1.41.1 tracking issue

### ISSUE THREAD

**Issue by dr-orlovsky** [rust-bitcoin/rust-bitcoin#510]
I wonder if we should have a Rust 2018 tracking issue and start tagging things that are blocked on it. 
        Would be good to have a clear picture of what we're missing by being on 1.29.

— @apoelstra https://github.com/rust-bitcoin/rust-bitcoin/pull/373#issuecomment-708395995

* No async support (not sure we need it at the level of this library)
* We can not rename packages, which creates a lot of ambiguity with `serde` crate/feature and prevents from using normal solutions like https://github.com/rust-bitcoin/rust-bitcoin/pull/373#issuecomment-689172390
* No able to use simple derive proc macros with `syn` 1.0 support like in [`amplify_derive`](https://crates.io/crates/amplify_derive) instead if hard-to-read `impl_` macro and multiple `impl From...` and `impl Display...` blocks (and there are wrapper types as well https://docs.rs/amplify_derive/2.0.6/amplify_derive/derive.Wrapper.html)
* Not able to do `#[non_exhaustive]` to solve the problem of breaking API changes with gated enum veriants

## Checklist

- [-] ~async - probably only for `Read` and `Write` async equivalents~ This is a nice to have, leave for another day.
- [x] revive #373 - WIP in PR #1006
- [-] ~use `amplify_derive`? note that this will blow up compile times for non-serde uses~ This is a nice to have, leave for another day.
- [x] Implement `TryFrom` and use `TryInto` (#1007 does `TryFrom` but not the `TryInto` bit)
- [x] `sort_by_cached_key` ~#524~ #1084
- [-] ~Use `alloc`? Is this still relevant?~
- [x] Use `Infallible` (#1075)
- [x] Use `contains()` on ranges https://github.com/rust-bitcoin/rust-bitcoin/pull/644#discussion_r707950661 (#952)
- [x] Use `chunks_exact()` instead of `chunks` (#952)
- [x] `use Trait as _` (#965)
- [-] ~Exhaustive integer matching instead of [comparing with zero](https://github.com/rust-bitcoin/rust-bitcoin/blob/615a900c88f1128b124fda76d821372bca61f796/src/util/amount.rs#L220)~ (improves readability) (#1012)
- [x] Use iter.copied() in https://github.com/rust-bitcoin/rust-bitcoin/pull/776#discussion_r784388347 (#1004)
- [x] Use `(;)?` instead of `(;)*` in macros (#716) (#1010)
- [x] Use `literal` instead of `expr` in macros (#716) (#1014)
- [x] Use`trim_start_matches` instead of `trim_left_matches` (#952)
- [x] Use `Vec::from([foo])` instead of `Vec::new` followed by `v.push()`. (#952, I checked all instances of `mut foo = Vec::new()` and also the same for `BTreeMap`)
- [x] Change `cause()` to `source()` in impl of `std::error::Error`
- [x] Use `u16::to_be_bytes` instead of `swap_bytes` (#952) 
- [x] Use `#[non_exhaustive]` (#1011 and #1026)

### `non_exhaustive` types:

- [x] `Error` (#1026)
- [x] `AddressType` (#1011)
- [x] `address::Payload` (#1011)
- [x] `LeafVersion` (#1011)

## PR list:
- [-] ~#635~ closed in favour of 983
- [x] https://github.com/rust-bitcoin/rust-bitcoin/pull/983

**Comment by apoelstra** [comment]
* `TryInto`, see https://github.com/rust-bitcoin/rust-bitcoin/pull/498 and probably many othercs

**Comment by RCasatta** [comment]
* sort_by_cached_key, see https://github.com/rust-bitcoin/rust-bitcoin/pull/524 (Rust 1.34)

**Comment by dr-orlovsky** [comment]
* stable `alloc` crate requiring 1.36 https://github.com/rust-bitcoin/rust-bitcoin/pull/528

**Comment by stevenroose** [comment]
- CommandString::try_into (see CommandString refactor as part of https://github.com/rust-bitcoin/rust-bitcoin/pull/496) can use `TryInto` (Rust 1.34)

**Comment by apoelstra** [comment]
1.36 is also the (current, may change without major version bump) MSRV of `rand` 0.8, whose non-optional dep tree now consists of `rand-core` and nothing else.

**Comment by apoelstra** [comment]
Possibly this crate could be interesting for us https://crates.io/crates/standback

**Comment by dpc** [comment]
mrust can compile rustc 1.39 now. https://www.reddit.com/r/rust/comments/mjxbaz/mrustc_upgrade_rustc_1390/

**Comment by elichai** [comment]
> mrust can compile rustc 1.39 now. https://www.reddit.com/r/rust/comments/mjxbaz/mrustc_upgrade_rustc_1390/

Nice!

**Comment by apoelstra** [comment]
Any opposition to moving to 1.39 then for rust-bitcoin 0.27? I can open another discussion issue. cc @TheBlueMatt 1.29 is really old now and is causing us a lot of grief especially with respect to nostd/`alloc`.

**Comment by TheBlueMatt** [comment]
Hmm, in general I think I'm fine with 1.39. It is less than two years old, though, so not sure if we don't want to use 1.34 (which is about two years old) or 1.36 (for alloc). I do admit it would be cool to go right to 1.39 and use async in a few places.

**Comment by apoelstra** [comment]
I think we could wait for the beginning of July and then jump to 1.36 (which aside from being "the one with `alloc`" also felt to me like a more important/well-tested release than 1.34 or 1.39).

**Comment by apoelstra** [comment]
As for using async I think that'd be a major change, though I guess if it were localized maybe not. And the upcoming API overhaul + transition to Rust 2018 would already be major changes, and it might be good to stagger them.

**Comment by TheBlueMatt** [comment]
Right, I was mostly thinking of (optionally) using async in rust-bitcoincore-rpc  by porting the lightning-block-sync stuff to there.

**Comment by elichai** [comment]
yeah I don't think rust-bitcoin itself should be async, maybe when we have AsyncRead/AsyncWrite in libstd then we might want to add support for these in the Encode/Decode traits

**Comment by elichai** [comment]
As for bumping, I can come up with a summary of what features we might want and what versions they exists in,
the first things that come to my mind are mostly dependencies (latest rand: 1.36, cc: 1.34, serde_derive: 1.31)

**Comment by TheBlueMatt** [comment]
Alright, lets plan on 1.36 Soon (tm), then. We may actually go ahead and do it in RL because we're already at 1.30 (due to cargo bug that prevented us from using 1.29) and I'm tired of trying to make cc work right.

**Comment by dr-orlovsky** [comment]
Upgrade to 1.36 actually means we can use `syn` with its MSRV 0.31, and finally do derives instead of macros for many occasions. 

I just completed development of a 1.31-compatible library simplifying creation of complex derivation macros to a big degree: https://docs.rs/amplify_syn/. You can see a sample of what it can be done with it here: https://github.com/LNP-BP/rust-amplify/blob/master/derive/src/lib.rs#L480-L553

Upon MSRV increase I will be happy to refactor rust-bitcoin code to introduce derivation macros using it (we already had one attempt in the past by @elichai, but then we were blocked by MSRV requirement).

**Comment by TheBlueMatt** [comment]
As of July 4 1.36 was two years old, so I think we can do this whenever now. https://blog.rust-lang.org/2019/07/04/Rust-1.36.0.html

**Comment by Kixunil** [comment]
One more thing to consider: use `base64` (MSRV 1.36) instead of `base64-compat` because mot libraries use that one anyway so we don't needlessly duplicate dependencies.

**Comment by Kixunil** [comment]
Let's compile a checklist:

- [ ] async - probably only for `Read` and `Write` async equivalents
- [ ] revive #373 
- [ ] use `amplify_derive`? note that this will blow up compile times for non-serde uses
- [ ] Use `#[non_exhaustive]`
- [ ] Implement `TryFrom` and use `TryInto`
- [ ] `sort_by_cached_key` #524 
- [ ] Use `alloc`? Is this still relevant?
- [ ] Use `Infallible`

**Comment by apoelstra** [comment]
With 1.34 apparently we get `convert::Infallible` which is pretty cool, see https://github.com/rust-bitcoin/rust-bech32/pull/43/

**Comment by Kixunil** [comment]
Already in my list :)

**Comment by tcharding** [comment]
@Kixunil can you add the following to your list above please
- Remove usage of `use Hash as HashTrait`, replace with `impl crate::Hash for Foo` and `use Hash as _`.

Discovered over here: https://github.com/rust-bitcoin/bitcoin_hashes/issues/16

**Comment by Kixunil** [comment]
Not a huge fan of it but added anyway. We can discuss further when it's relevant.

**Comment by RCasatta** [comment]
https://github.com/rust-bitcoin/rust-bitcoin/pull/638 use `rustdoc::broken_intra_doc_links`

**Comment by Kixunil** [comment]
@RCasatta that one is even later, I think

**Comment by tcharding** [comment]
Could this issue be better titled `Bump MSRV tracking issue`?

**Comment by tcharding** [comment]
On January 30 Rust 1.41 will be two years old.
https://blog.rust-lang.org/2020/01/30/Rust-1.41.0.html

**Comment by Kixunil** [comment]
I would've liked 1.41.1 ((Feb 27)[https://blog.rust-lang.org/2020/02/27/Rust-1.41.1.html]) more because of fixed issues and not being super-ancient. (It's in Debian oldstable)

**Comment by apoelstra** [comment]
I'd be ok with 1.41. It's only two versions ahead of mrustc support.

**Comment by tcharding** [comment]
+1 for Rust `1.41.1`, with the suggestion by the Rust team to not use 1.4.0 seems that we should specify the dot release not just `1.41`

**Comment by tcharding** [comment]
With Taproot release imminent, I've been spending some time thinking through the MSRV update across the board. PRs for [rust-secp256k1](https://github.com/rust-bitcoin/rust-secp256k1/pull/331) and [rust-bitcoin](https://github.com/rust-bitcoin/rust-bitcoin/pull/635) were already done by @elichai and @Kixunil respectively. I hope I am not [treading on your toes](https://www.merriam-webster.com/dictionary/tread%20on%20the%20toes%20of%20%28someone%29). I've pushed branches to my own tree for both with changes to make them fit in with what I propose the mantra for this biggish change should be - "minimum code changes to get the code to build with edition 2018 enabled". We can do a bunch of follow ups, it is going to be hard enough to co-ordinate updating all the crates together without adding any additional ~bikeshedding~ reviewing.

Note:
- Clippy changes are equired when clippy runs in CI (bech32 and bitcoinconcensus) - these PRs will need squashing.
- Removal of leading `::` is not strictly inline with the mantra but fits in with fixing import statements so its included.

PR list
- [rust-bech32](https://github.com/rust-bitcoin/rust-bech32/pull/57)
- [rust-bitcoinconcensus](https://github.com/rust-bitcoin/rust-bitcoinconsensus/pull/34)
- [bitcoin_hashes](https://github.com/rust-bitcoin/bitcoin_hashes/pull/136)
- [rust-secp256k1](https://github.com/rust-bitcoin/rust-secp256k1/pull/331)
- [rust-bitcoin](https://github.com/rust-bitcoin/rust-bitcoin/pull/635)

PRs from above with only minimal changes (links to my tree):
- [rust-secp256k1](https://github.com/tcharding/rust-secp256k1/tree/edition-2018)
- [rust-bitcoin](https://github.com/tcharding/rust-bitcoin/tree/edition-2018)

Followups (containing the work of @elichain and @Kikunil
- [rust-secp256k1](https://github.com/tcharding/rust-secp256k1/tree/edition-2018-followups)
- [rust-bitcoin](https://github.com/tcharding/rust-bitcoin/tree/edition-2018-followups)

**Comment by apoelstra** [comment]
Definitely not treading on *my* toes. Thanks so much for doing this!

cc @TheBlueMatt are you okay with
* Bumping MSRV to rustc 1.41.1, which will be 2 years old in a couple weeks, and is two versions ahead of mrustc's 1.39 support, and is in Debian stable (though only for popular arches)
* Doing an ecosystem-wide MSRV bump as the next project after the rust-bitcoin Taproot release

**Comment by dr-orlovsky** [comment]
+1 for 1.41.1

**Comment by TheBlueMatt** [comment]
Oops, sorry I'd missed this before. I'm okay with 1.41.1, though given where mrustc is we may end up kinda stuck there for some time :/

**Comment by Kixunil** [comment]
I've moved the checklist to the issue description to be easier to find. Hope @dr-orlovsky doesn't mind...

**Comment by Kixunil** [comment]
>  though given where mrustc is we may end up kinda stuck there for some time :/

I don't think so, [mrustc allegedly supports 1.54](https://github.com/thepowersgang/mrustc/blob/master/build-1.54.0.sh)

**Comment by TheBlueMatt** [comment]
Nice, that's pertty new! Their readme still says 1.39.0, but I guess they're working on more.

**Comment by junderw** [comment]
+1 for 1.41.1

**Comment by dpc** [comment]
Relevant: https://www.reddit.com/r/rust/comments/ttaqay/progress_report_10_rustc_codegen_gcc_can_now/

>  As you may have guessed from the title of this blog post, the GCC codegen has made enough progress to be able to compile rustc itself! That’s a huge milestone!

**Comment by sanket1729** [comment]
We should start working on this now that we have 0.28.0

**Comment by Kixunil** [comment]
Can we first get existing PRs merged? At least mine. :)

**Comment by sanket1729** [comment]
Sorry for missing this. Working on reviewing it

**Comment by sanket1729** [comment]
Thanks for creating the tag. Reviewers can be `before MSRV bump` tag to see which PRs to review before MSRV bump. 

I have no opposition to merging those before the MSRV update, I am wondering why those must be done before the MSRV update? @Kixunil

**Comment by Kixunil** [comment]
It'll be less churn merging existing PRs and then changing editions because changing edition is mostly automated process. See https://github.com/rust-bitcoin/rust-bitcoin/discussions/945

**Comment by tcharding** [comment]
To save anyone else trying to grep for things on the checklist like I've been doing for the last two hours - clippy will find many of these for us :)

**Comment by Kixunil** [comment]
Renamed the tag to "before *edition* bump" as it better conveys the idea. IOW we can absolutely merge #964 before these.

**Comment by tcharding** [comment]
Related 'Discusssions' thread: https://github.com/rust-bitcoin/rust-bitcoin/discussions/945

**Comment by tcharding** [comment]
I've checked off all the basic ones crew, many of the others I either do not understand how to do or I do not understand the motivation behind doing them. If you put one on the list, and would like to see it implemented without having to do it yourself, please throw a comment here with a bit more context and I'm happy to do it. Thanks.

**Comment by Kixunil** [comment]
Adding `#[non_exhaustive]` attribute to the listed types would be great. It ensures forward compatibility.

`$(,)?` in macros is cosmetic.

**Comment by tcharding** [comment]
I had a think about the `non_exhaustive` task. I have a question: Why only do `Error`, if we are justified in doing one error why not do all of the error types?

**Comment by tcharding** [comment]
Does anyone remember where `Infallible` was supposed to be used?

**Comment by tcharding** [comment]
I scratched `alloc` from the list, flagging it here because I'm not 100% sure that is correct.

**Comment by Kixunil** [comment]
@tcharding I think by `Error` I meant all error types. :)

```
$ rg --multiline 'enum.*\{[ \t\n]*\}'
src/blockdata/script.rs
173:enum Uninhabited {}
```

`Uninhabited` should be removed and replaced with `Infallible`

We can skip `alloc` because the library is mostly useless without it and it'd be easier to do with `Amount` & co in a separate crate.

**Comment by tcharding** [comment]
Cool, cheers.

**Comment by tcharding** [comment]
There are only a few things to go on the list.
- The `Infallible`/`Uninhabited` thing
- The two checklist items at the top that do not have PRs, neither of which I'm know if we still want, anyone?
  - " async - probably only for Read and Write async equivalents"
  - "use amplify_derive? note that this will blow up compile times for non-serde uses"

**Comment by dr-orlovsky** [comment]
[`amplify_derive`](https://github.com/rust-amplify/rust-amplify/tree/master/derive) is a crate which I did to simplify deriving [`Error`s](https://docs.rs/amplify_derive/2.11.2/amplify_derive/derive.Error.html) and [`Display`s](https://docs.rs/amplify_derive/2.11.2/amplify_derive/derive.Display.html) all around, which may be useful here. That crate has very few dependencies (comparing to other error/display derive crates), and matches MSRV after its bump. So it's up to you to decide whether to use it here or continue maintaining larger codebase for each error.

Me and @Kixunil are maintainers of that repo (he contributed to other parts of the `amplify` lib family) and I can add more maintainers from here to ensure safety of the use of the library & reviews requirement for any codebase changes.

**Comment by dr-orlovsky** [comment]
async - that was about adding support to async functions, for instance for implementing non-blocking consensus encoding/decoding with `io::Read/Write` - this may be useful, for instance when they are used in serializing large structures (blocks) to files by projects like electrum server (personally I do not need it, since I am using different approach to a non-blocking architecture with ZMQ and microservices - and not rust `async` which I do not like).

**Comment by Kixunil** [comment]
Exactly what @dr-orlovsky said. Those two are definitely not required for the issue to be complete and are more "nice to have" things that could be added at any time in the future. IMO it's even better to skip them to release more quickly.

@tcharding I belive we already did `Infallible`. I checked also wrong uses of `()` and while there is one in `AddressType` it's not infallible, just a bad error type.
